_posts/2020-07-15-how-to-bring-your-own-codegen-to-tvm.md

```diff
@@ -0,0 +1,499 @@
+
+{:center: style="text-align: center"}
+![After Merging Compiler Regions](/images/bring-your-own-codegen/after_merging_regions.png){: width="50%"}
+{:center}
+<center>
+Figure 3: After Merging Compiler Regions.
+</center> <p></p>
+
+**2.2: Partition Graph**: For each region from the previous step, we create a Relay function with an attribute `Compiler` to indicate that this Relay function should be entirely offloaded to your accelerator, as shown in Figure 4.
+
+{:center: style="text-align: center"}
+![After Graph Partitioning](/images/bring-your-own-codegen/after_partitioning.png){: width="50%"}
+{:center}
+<center>
+Figure 4: After Graph Partitioning.
+</center> <p></p>
+
+### 3. Code Generation
+Now we know which part of the Relay graph should be offloaded. In this step, we sequentially send every Relay function with `Compiler=your_accelerator` to your codegen. Your codegen should compile the Relay function to the form that matches your own compilation flow. It can be either C source code or any text formats.
+
+Finally, all compiled functions will be serialized along with other non-offloaded Relay functions to a single `.so` file by the TVM `export_library` Python API. In other words, the user will get only one `.so` file after running this flow.
+
+### 4. Runtime
+You may also need to implement a runtime to initialize your graph engine (if applicable) and execute the compiled functions. During the inference, TVM runtime (i.e., graph runtime or VM) will leverage your runtime to invoke the offloaded functions when the TVM runtime encounters the corresponding function call in Figure 4. Your runtime is responsible for launching the compiled function with the given input tensor arrays and filling in the results to the output tensor arrays.
+
+In the rest of this post, we use DNNL as an example to demonstrate how to achieve the above workflow using the BYOC framework. Please note that all referred code and line number in this post are based on the TVM repository's master branch commit [8a0249c](https://github.com/apache/incubator-tvm/tree/8a0249cd4d12a2eb1a4e7a692a9265bc63fec5c8).
+
+## Bring DNNL to TVM: Annotation Rules
+
+The BYOC framework provides two approaches for you to describe the supported operators and patterns. You can use both of them simultaneously. In this section, we use DNNL as an example to show how to make use of them. The complete implementation is available [here](https://github.com/apache/incubator-tvm/blob/8a0249cd4d12a2eb1a4e7a692a9265bc63fec5c8/python/tvm/relay/op/contrib/dnnl.py). Note that we put the annotation rules for your codegen under `python/tvm/relay/op/contrib/your_codegen_name.py`.
```

Comment:
Not directly a point on this blog post, but it's made me realise the directory where we put all the rules seems a bit arbitrary and unintuitive. Maybe we could consider making a BYOC directory under tvm/relay?

Reply:
We currently put everything under differernt `contrib` for different purposes:
`python/tvm/relay/op/contrib/your_codegen_name.py`: The annotation rules in Python.
`src/relay/backend/contrib/your_codegen_name/*`: The codegen in C++.
`src/runtime/contrib/your_codegen_name/*`: The runtime in C++.

Per your suggestion, the C++ parts might be moved to `src/relay/byoc/your_name/*`, but the Python part has to be in a separate place anyways. We can discuss this issue along with the codegen specific transform sequence APIs in the discuss forum later on.